worksheet_B/Mandelbrot.cpp

```diff
@@ -12,44 +90,133 @@ int main()
+			double localPositionX = mousePosition.x - display._window_x;
+			double localPositionY = mousePosition.y - display._window_y;
+
+			// Compare against window dimensions to determine mouse cursor position in terms of data bounds
+			double ratioX = localPositionX / display._window_width;
+			double dataRangeCursorPositionX = (ratioX * currentSideLength) + minX;
+
+			double ratioY = localPositionY / display._window_height;
+			double dataRangeCursorPositionY = (ratioY * currentSideLength) + minY;
+
+			// Apply zoom modifier to alter currentSideLength
+			if (zoomIn)
+				currentSideLength /= zoomMagnitude;
+			else
+				currentSideLength *= zoomMagnitude;
+
+			// Determine new minX, minY values for zoomed section with mouse cursor as new centre
+			minX = dataRangeCursorPositionX - (currentSideLength / 2);
+			minY = dataRangeCursorPositionY - (currentSideLength / 2);
+
+			// Determine new maxX, maxY values
+			maxX = minX + currentSideLength;
+			maxY = minY + currentSideLength;
+
+			// Display zoom level to console
+			double zoomLevel = originalSideLength / currentSideLength;
+			std::cout << "Zoom magnitude: " << zoomLevel << std::endl;
+		}
+
+			// Generate the image
+		for (double pixelY = 0; pixelY < image.height(); pixelY++)
 		{
-			// TODO: Map the x coordinate into the range minX to maxX
-			//double x0 =
+			// Map the y coordinate into the range minY to maxY
+			double y0 = (pixelY / image.height()) * (maxY - minY) + minY;
+
+			for (double pixelX = 0; pixelX < image.width(); pixelX++)
+			{
+				// Map the x coordinate into the range minX to maxX
+				double x0 = (pixelX / image.width()) * (maxX - minX) + minX;
 
-			// TODO: implement the algorithm to colour a single pixel (x0, y0) of the Mandelbrot set fractal
-			// The code below simply fills the screen with random pixels
+				// Algorithm to colour a single pixel (x0, y0) of the Mandelbrot set fractal
 
-			// Write the pixel
-			// TODO: change the right-hand side of these three lines to write the desired pixel colour value
-			image(pixelX, pixelY, 0, 0) = rand(); // red component
-			image(pixelX, pixelY, 0, 1) = rand(); // green component
-			image(pixelX, pixelY, 0, 2) = rand(); // blue component
+				double x = 0;
+				double y = 0;
+				int i = 0;
+
+				while (x*x + y*y < 4 && i < 200)
+				{
+					double xTemp = x*x - y*y + x0;
+					y = (2.0 * x * y) + y0;
+					x = xTemp;
+					i++;
+				}
+
+				// Write the pixel
+				rgb currentRGB;
+
+				if (i >= 200)										// Upper limit of 200 to catch outliers
+				{
+					currentRGB = { 0.0, 0.0, 0.0 };					// Black specified for outlying values of i
+				}
+				else
+				{
+					hsv currentHSV = { i, 1.0, 1.0 };				// Magic numbers for saturation and value implicit in worksheet outline
+					currentRGB = hsv2rgb(currentHSV);
+				}
+
+				image(pixelX, pixelY, 0, 0) = currentRGB.r * 255.0;	// Don't know why the hsv2rgb() function wouldn't do this?
```

Comment:
There are two standard ways of handling RGB components: as integers between 0 and 255, or as floating point numbers between 0 and 1. It's quite common to have to convert from one to the other.